README
NAME

    Archive::Libarchive::FFI - Perl bindings to libarchive via FFI

VERSION

    version 0.0802

SYNOPSIS

    list archive filenames

     use Archive::Libarchive::FFI qw( :all );
     
     my $archive = archive_read_new();
     archive_read_support_filter_all($archive);
     archive_read_support_format_all($archive);
     # example is a tar file, but any supported format should work
     # (zip, iso9660, etc.)
     archive_read_open_filename($archive, 'archive.tar', 10240);
     
     while(archive_read_next_header($archive, my $entry) == ARCHIVE_OK)
     {
       print archive_entry_pathname($entry), "\n";
       archive_read_data_skip($archive);
     }
     
     archive_read_free($archive);

    extract archive

     use Archive::Libarchive::FFI qw( :all );
     
     my $archive = archive_read_new();
     archive_read_support_filter_all($archive);
     archive_read_support_format_all($archive);
     my $disk = archive_write_disk_new();
     archive_write_disk_set_options($disk, 
       ARCHIVE_EXTRACT_TIME   |
       ARCHIVE_EXTRACT_PERM   |
       ARCHIVE_EXTRACT_ACL    |
       ARCHIVE_EXTRACT_FFLAGS
     );
     archive_write_disk_set_standard_lookup($disk);
     archive_read_open_filename($archive, 'archive.tar', 10240);
     
     while(1)
     {
       my $r = archive_read_next_header($archive, my $entry);
       last if $r == ARCHIVE_EOF;
       
       archive_write_header($disk, $entry);
       
       while(1)
       {
         my $r = archive_read_data_block($archive, my $buffer, my $offset);
         last if $r == ARCHIVE_EOF;
         archive_write_data_block($disk, $buffer, $offset);
       }
     }
     
     archive_read_close($archive);
     archive_read_free($archive);
     archive_write_close($disk);
     archive_write_free($disk);

    write archive

     use File::stat;
     use File::Slurp qw( read_file );
     use Archive::Libarchive::FFI qw( :all );
     
     my $archive = archive_write_new();
     # many other formats are supported ...
     archive_write_set_format_pax_restricted($archive);
     archive_write_open_filename($archive, 'archive.tar');
     
     foreach my $filename (@filenames)
     {
       my $entry = archive_entry_new();
       archive_entry_set_pathname($entry, $filename);
       archive_entry_set_size($entry, stat($filename)->size);
       archive_entry_set_filetype($entry, AE_IFREG);
       archive_entry_set_perm($entry, 0644);
       archive_write_header($archive, $entry);
       archive_write_data($archive, scalar read_file($filename));
       archive_entry_free($entry);
     }
     archive_write_close($archive);
     archive_write_free($archive);

DESCRIPTION

    This module provides a functional interface to libarchive. libarchive
    is a C library that can read and write archives in a variety of formats
    and with a variety of compression filters, optimized in a stream
    oriented way. A familiarity with the libarchive documentation would be
    helpful, but may not be necessary for simple tasks. The documentation
    for this module is split into four separate documents:

    Archive::Libarchive::FFI

      This document, contains an overview and some examples.

    Archive::Libarchive::FFI::Callback

      Documents the callback interface, used for customizing input and
      output.

    Archive::Libarchive::FFI::Constant

      Documents the constants provided by this module.

    Archive::Libarchive::FFI::Function

      The function reference, includes a list of all functions provided by
      this module.

    If you are linking against an older version of libarchive, some
    functions and constants may not be available. You can use the can
    method to test if a function or constant is available, for example:

     if(Archive::Libarchive::FFI->can('archive_read_support_filter_grzip')
     {
       # grzip filter is available.
     }
     
     if(Archive::Libarchive::FFI->can('ARCHIVE_OK'))
     {
       # ... although ARCHIVE_OK should always be available.
     }

EXAMPLES

    These examples are translated from equivalent C versions provided on
    the libarchive website, and are annotated here with Perl specific
    details. These examples are also included with the distribution.

 List contents of archive stored in file

     use strict;
     use warnings;
     use Archive::Libarchive::FFI qw( :all );
     
     # this is a translation to perl for this:
     #  https://github.com/libarchive/libarchive/wiki/Examples#wiki-List_contents_of_Archive_stored_in_File
     
     my $a = archive_read_new();
     archive_read_support_filter_all($a);
     archive_read_support_format_all($a);
     
     my $r = archive_read_open_filename($a, "archive.tar", 10240);
     if($r != ARCHIVE_OK)
     {
       die "error opening archive.tar: ", archive_error_string($a);
     }
     
     while (archive_read_next_header($a, my $entry) == ARCHIVE_OK)
     {
       print archive_entry_pathname($entry), "\n";
       archive_read_data_skip($a); 
     }
     
     $r = archive_read_free($a);
     if($r != ARCHIVE_OK)
     {
       die "error freeing archive";
     }

 List contents of archive stored in memory

     use strict;
     use warnings;
     use Archive::Libarchive::FFI qw( :all );
     
     # this is a translation to perl for this:
     #  https://github.com/libarchive/libarchive/wiki/Examples#wiki-List_contents_of_Archive_stored_in_Memory
     
     my $buff = do {
       open my $fh, '<', "archive.tar.gz";
       local $/;
       <$fh>
     };
     
     my $a = archive_read_new();
     archive_read_support_filter_gzip($a);
     archive_read_support_format_tar($a);
     my $r = archive_read_open_memory($a, $buff);
     if($r != ARCHIVE_OK)
     {
       print "r = $r\n";
       die "error opening archive.tar: ", archive_error_string($a);
     }
     
     while (archive_read_next_header($a, my $entry) == ARCHIVE_OK) {
       print archive_entry_pathname($entry), "\n";
       archive_read_data_skip($a); 
     }
     
     $r = archive_read_free($a);
     if($r != ARCHIVE_OK)
     {
       die "error freeing archive";
     }

 List contents of archive with custom read functions

     use strict;
     use warnings;
     use Archive::Libarchive::FFI qw( :all );
     
     list_archive(shift @ARGV);
     
     sub list_archive
     {
       my $name = shift;
       my %mydata;
       my $a = archive_read_new();
       $mydata{name} = $name;
       open $mydata{fh}, '<', $name;
       archive_read_support_filter_all($a);
       archive_read_support_format_all($a);
       archive_read_open($a, \%mydata, undef, \&myread, \&myclose);
       while(archive_read_next_header($a, my $entry) == ARCHIVE_OK)
       {
         print archive_entry_pathname($entry), "\n";
       }
       archive_read_free($a);
     }
     
     sub myread
     {
       my($archive, $mydata) = @_;
       my $br = read $mydata->{fh}, my $buffer, 10240;
       return (ARCHIVE_OK, $buffer);
     }
     
     sub myclose
     {
       my($archive, $mydata) = @_;
       close $mydata->{fh};
       %$mydata = ();
       return ARCHIVE_OK;
     }

 A universal decompressor

     use strict;
     use warnings;
     use Archive::Libarchive::FFI qw( :all );
     
     # this is a translation to perl for this:
     #  https://github.com/libarchive/libarchive/wiki/Examples#a-universal-decompressor
     
     my $r;
     
     my $a = archive_read_new();
     archive_read_support_filter_all($a);
     archive_read_support_format_raw($a);
     $r = archive_read_open_filename($a, "hello.txt.gz.uu", 16384);
     if($r != ARCHIVE_OK)
     {
       die archive_error_string($a);
     }
     
     $r = archive_read_next_header($a, my $ae);
     if($r != ARCHIVE_OK)
     {
       die archive_error_string($a);     
     }
     
     while(1)
     {
       my $size = archive_read_data($a, my $buff, 1024);
       if($size < 0)
       {
         die archive_error_string($a);
       }
       if($size == 0)
       {
         last;
       }
       print $buff;
     }
     
     archive_read_free($a);

 A basic write example

     use strict;
     use warnings;
     use autodie;
     use File::stat;
     use Archive::Libarchive::FFI qw( :all );
     
     # this is a translation to perl for this:
     #  https://github.com/libarchive/libarchive/wiki/Examples#wiki-A_Basic_Write_Example
     
     sub write_archive
     {
       my($outname, @filenames) = @_;
       
       my $a = archive_write_new();
       
       archive_write_add_filter_gzip($a);
       archive_write_set_format_pax_restricted($a);
       archive_write_open_filename($a, $outname);
       
       foreach my $filename (@filenames)
       {
         my $st = stat $filename;
         my $entry = archive_entry_new();
         archive_entry_set_pathname($entry, $filename);
         archive_entry_set_size($entry, $st->size);
         archive_entry_set_filetype($entry, AE_IFREG);
         archive_entry_set_perm($entry, 0644);
         archive_write_header($a, $entry);
         open my $fh, '<', $filename;
         my $len = read $fh, my $buff, 8192;
         while($len > 0)
         {
           archive_write_data($a, $buff);
           $len = read $fh, $buff, 8192;
         }
         close $fh;
         
         archive_entry_free($entry);
       }
       archive_write_close($a);
       archive_write_free($a);
     }
     
     unless(@ARGV > 0)
     {
       print "usage: perl basic_write.pl archive.tar.gz file1 [ file2 [ ... ] ]\n";
       exit 2;
     }
     
     unless(@ARGV > 1)
     {
       print "Cowardly refusing to create an empty archive\n";
       exit 2;
     }
     
     write_archive(@ARGV);

 Constructing objects on disk

     use strict;
     use warnings;
     use Archive::Libarchive::FFI qw( :all );
     
     # this is a translation to perl for this:
     #  https://github.com/libarchive/libarchive/wiki/Examples#wiki-Constructing_Objects_On_Disk
     
     my $a = archive_write_disk_new();
     archive_write_disk_set_options($a, ARCHIVE_EXTRACT_TIME);
     
     my $entry = archive_entry_new();
     archive_entry_set_pathname($entry, "my_file.txt");
     archive_entry_set_filetype($entry, AE_IFREG);
     archive_entry_set_size($entry, 5);
     archive_entry_set_mtime($entry, 123456789, 0);
     archive_entry_set_perm($entry, 0644);
     archive_write_header($a, $entry);
     archive_write_data($a, "abcde");
     archive_write_finish_entry($a);
     archive_write_free($a);
     archive_entry_free($entry);

 A complete extractor

     use strict;
     use warnings;
     use Archive::Libarchive::FFI qw( :all );
     
     # this is a translation to perl for this:
     #  https://github.com/libarchive/libarchive/wiki/Examples#wiki-A_Complete_Extractor
     
     my $filename = shift @ARGV;
     
     unless(defined $filename)
     {
       warn "reading archive from standard in";
     }
     
     my $r;
     
     my $flags = ARCHIVE_EXTRACT_TIME
               | ARCHIVE_EXTRACT_PERM
               | ARCHIVE_EXTRACT_ACL
               | ARCHIVE_EXTRACT_FFLAGS;
     
     my $a = archive_read_new();
     archive_read_support_filter_all($a);
     archive_read_support_format_all($a);
     my $ext = archive_write_disk_new();
     archive_write_disk_set_options($ext, $flags);
     archive_write_disk_set_standard_lookup($ext);
     
     $r = archive_read_open_filename($a, $filename, 10240);
     if($r != ARCHIVE_OK)
     {
       die "error opening $filename: ", archive_error_string($a);
     }
     
     while(1)
     {
       $r = archive_read_next_header($a, my $entry);
       if($r == ARCHIVE_EOF)
       {
         last;
       }
       if($r != ARCHIVE_OK)
       {
         print archive_error_string($a), "\n";
       }
       if($r < ARCHIVE_WARN)
       {
         exit 1;
       }
       $r = archive_write_header($ext, $entry);
       if($r != ARCHIVE_OK)
       {
         print archive_error_string($ext), "\n";
       }
       elsif(archive_entry_size($entry) > 0)
       {
         copy_data($a, $ext);
       }
     }
     
     archive_read_close($a);
     archive_read_free($a);
     archive_write_close($ext);
     archive_write_free($ext);
     
     sub copy_data
     {
       my($ar, $aw) = @_;
       my $r;
       while(1)
       {
         $r = archive_read_data_block($ar, my $buff, my $offset);
         if($r == ARCHIVE_EOF)
         {
           return;
         }
         if($r != ARCHIVE_OK)
         {
           die archive_error_string($ar), "\n";
         }
         $r = archive_write_data_block($aw, $buff, $offset);
         if($r != ARCHIVE_OK)
         {
           die archive_error_string($aw), "\n";
         }
       }
     }

 Unicode

    Libarchive deals with two types of string like data. Pathnames, user
    and group names are proper strings and are encoded in the codeset for
    the current POSIX locale. Content data for files stored and retrieved
    from in raw bytes.

    The usual operational procedure in Perl is to convert everything on
    input into UTF-8, operate on the UTF-8 data and then convert (if
    necessary) everything on output to the desired output format.

    In order to get useful string data out of libarchive, this module
    translates its input/output using the codeset for the current POSIX
    locale. So you must be using a POSIX locale that supports the
    characters in the pathnames of the archives you are going to process,
    and it is highly recommend that you use a UTF-8 locale, which should
    cover everything.

     use strict;
     use warnings;
     use utf8;
     use Archive::Libarchive::FFI qw( :all );
     use POSIX qw( setlocale LC_ALL );
     
     # substitute en_US.utf8 for the correct UTF-8 locale for your region.
     setlocale(LC_ALL, "en_US.utf8"); # or 'export LANG=en_US.utf8' from your shell.
     
     my $entry = archive_entry_new();
     
     archive_entry_set_pathname($entry, "привет.txt");
     my $string = archive_entry_pathname($entry); # "привет.txt"
     
     archive_entry_free($entry);

    If you try to pass a string with characters unsupported by your current
    locale, the behavior is undefined. If you try to retrieve strings with
    characters unsupported by your current locale you will get undef.

    Unfortunately locale names are not portable across systems, so you
    should probably not hard code the locale as shown here unless you know
    the correct locale name for all the platforms that your script will
    run.


Changes
Revision history for Archive-Libarchive-FFI

0.0802    2014-06-19 07:48:01 -0400
  - promote to production

0.0801_01 2014-06-18 15:55:44 -0400
  - use Alien::Libarchive::Installer if possible
    instead of the heavier weight Alien::Libarchive

0.0801    2014-06-10 15:40:01 -0400
  - Fix bundled dists

0.0800    2014-06-10 12:54:29 -0400
  - Archive::Libarchive 0.19 compatability

0.0706    2014-03-31 11:10:49 -0400
  - fix for OpenBSD

0.0705    2014-03-27 12:59:38 -0400
  - for libarchive 2.x, use our own Perl implementation of archive_read_open_memory
    as interfacing with the C version is causing SIGSEGV.

0.0704    2014-03-27 10:41:31 -0400
  - Take advantage of FFI::Raw 0.29's ability to return strings from callbacks
  - And therefore remove dependency on Module::Build::FFI

0.0703    2014-03-24 17:17:33 -0400
  - fixed some type declarations which were causing failures on 32bit Linux
    (possibly others)
    specifically: use int64_t instead of gid_t/uid_t for UID/GID types
                  use size_t instead of int64_t for buffer sizes

0.0702    2014-02-22 10:30:50 -0500
  - require FFI::Raw 0.28 and FFI::Util 0.06 for recent bugfixes
  
0.07      2014-02-10 18:54:14 -0500
  - fix MSWin32 build failure
  - support for libarchive 2.5.x
  - Update tests to be in sync with Archive::Libarchive::XS 0.07

0.06      2014-02-05 15:14:53 -0500
  - callback functions were made optional, as they are in the XS version
    (support for older versions of libarchive)
  - various fixes to support libarchive 2.x
    including versons of libarchive that do not have archive_read/write_free
    (use archive_read/write_finish instead)
  - fixed argument count bug that only manifested itself on FreeBSD (GH#13)
  - build fixes for NetBSD

0.05      2014-02-04 13:24:04 -0500
  - bump version to match Archive::Libarchive::XS

0.04      2014-02-04 13:22:23 -0500
  - firsty CPAN release


lib/Archive/Libarchive/FFI/Constant.pm
package Archive::Libarchive::FFI::Constant;

use strict;
use warnings;

our $VERSION = '0.0802'; # VERSION

package
  Archive::Libarchive::FFI;

use constant {
  AE_IFBLK => 24576,
  AE_IFCHR => 8192,
  AE_IFDIR => 16384,
  AE_IFIFO => 4096,
  AE_IFLNK => 40960,
  AE_IFMT => 61440,
  AE_IFREG => 32768,
  AE_IFSOCK => 49152,
  ARCHIVE_COMPRESSION_BZIP2 => 2,
  ARCHIVE_COMPRESSION_COMPRESS => 3,
  ARCHIVE_COMPRESSION_GZIP => 1,
  ARCHIVE_COMPRESSION_LRZIP => 10,
  ARCHIVE_COMPRESSION_LZIP => 9,
  ARCHIVE_COMPRESSION_LZMA => 5,
  ARCHIVE_COMPRESSION_NONE => 0,
  ARCHIVE_COMPRESSION_PROGRAM => 4,
  ARCHIVE_COMPRESSION_RPM => 8,
  ARCHIVE_COMPRESSION_UU => 7,
  ARCHIVE_COMPRESSION_XZ => 6,
  ARCHIVE_ENTRY_ACL_ADD_FILE => 16,
  ARCHIVE_ENTRY_ACL_ADD_SUBDIRECTORY => 32,
  ARCHIVE_ENTRY_ACL_APPEND_DATA => 32,
  ARCHIVE_ENTRY_ACL_DELETE => 2048,
  ARCHIVE_ENTRY_ACL_DELETE_CHILD => 256,
  ARCHIVE_ENTRY_ACL_ENTRY_DIRECTORY_INHERIT => 67108864,
  ARCHIVE_ENTRY_ACL_ENTRY_FAILED_ACCESS => 1073741824,
  ARCHIVE_ENTRY_ACL_ENTRY_FILE_INHERIT => 33554432,
  ARCHIVE_ENTRY_ACL_ENTRY_INHERIT_ONLY => 268435456,
  ARCHIVE_ENTRY_ACL_ENTRY_NO_PROPAGATE_INHERIT => 134217728,
  ARCHIVE_ENTRY_ACL_ENTRY_SUCCESSFUL_ACCESS => 536870912,
  ARCHIVE_ENTRY_ACL_EVERYONE => 10107,
  ARCHIVE_ENTRY_ACL_EXECUTE => 1,
  ARCHIVE_ENTRY_ACL_GROUP => 10003,
  ARCHIVE_ENTRY_ACL_GROUP_OBJ => 10004,
  ARCHIVE_ENTRY_ACL_INHERITANCE_NFS4 => 2113929216,
  ARCHIVE_ENTRY_ACL_LIST_DIRECTORY => 8,
  ARCHIVE_ENTRY_ACL_MASK => 10005,
  ARCHIVE_ENTRY_ACL_OTHER => 10006,
  ARCHIVE_ENTRY_ACL_PERMS_NFS4 => 65529,
  ARCHIVE_ENTRY_ACL_PERMS_POSIX1E => 7,
  ARCHIVE_ENTRY_ACL_READ => 4,
  ARCHIVE_ENTRY_ACL_READ_ACL => 4096,
  ARCHIVE_ENTRY_ACL_READ_ATTRIBUTES => 512,
  ARCHIVE_ENTRY_ACL_READ_DATA => 8,
  ARCHIVE_ENTRY_ACL_READ_NAMED_ATTRS => 64,
  ARCHIVE_ENTRY_ACL_STYLE_EXTRA_ID => 1024,
  ARCHIVE_ENTRY_ACL_STYLE_MARK_DEFAULT => 2048,
  ARCHIVE_ENTRY_ACL_SYNCHRONIZE => 32768,
  ARCHIVE_ENTRY_ACL_TYPE_ACCESS => 256,
  ARCHIVE_ENTRY_ACL_TYPE_ALARM => 8192,
  ARCHIVE_ENTRY_ACL_TYPE_ALLOW => 1024,
  ARCHIVE_ENTRY_ACL_TYPE_AUDIT => 4096,
  ARCHIVE_ENTRY_ACL_TYPE_DEFAULT => 512,
  ARCHIVE_ENTRY_ACL_TYPE_DENY => 2048,
  ARCHIVE_ENTRY_ACL_TYPE_NFS4 => 15360,
  ARCHIVE_ENTRY_ACL_TYPE_POSIX1E => 768,
  ARCHIVE_ENTRY_ACL_USER => 10001,
  ARCHIVE_ENTRY_ACL_USER_OBJ => 10002,
  ARCHIVE_ENTRY_ACL_WRITE => 2,
  ARCHIVE_ENTRY_ACL_WRITE_ACL => 8192,
  ARCHIVE_ENTRY_ACL_WRITE_ATTRIBUTES => 1024,
  ARCHIVE_ENTRY_ACL_WRITE_DATA => 16,
  ARCHIVE_ENTRY_ACL_WRITE_NAMED_ATTRS => 128,
  ARCHIVE_ENTRY_ACL_WRITE_OWNER => 16384,
  ARCHIVE_EOF => 1,
  ARCHIVE_EXTRACT_ACL => 32,
  ARCHIVE_EXTRACT_FFLAGS => 64,
  ARCHIVE_EXTRACT_HFS_COMPRESSION_FORCED => 32768,
  ARCHIVE_EXTRACT_MAC_METADATA => 8192,
  ARCHIVE_EXTRACT_NO_AUTODIR => 1024,
  ARCHIVE_EXTRACT_NO_HFS_COMPRESSION => 16384,
  ARCHIVE_EXTRACT_NO_OVERWRITE => 8,
  ARCHIVE_EXTRACT_NO_OVERWRITE_NEWER => 2048,
  ARCHIVE_EXTRACT_OWNER => 1,
  ARCHIVE_EXTRACT_PERM => 2,
  ARCHIVE_EXTRACT_SECURE_NODOTDOT => 512,
  ARCHIVE_EXTRACT_SECURE_SYMLINKS => 256,
  ARCHIVE_EXTRACT_SPARSE => 4096,
  ARCHIVE_EXTRACT_TIME => 4,
  ARCHIVE_EXTRACT_UNLINK => 16,
  ARCHIVE_EXTRACT_XATTR => 128,
  ARCHIVE_FAILED => -25,
  ARCHIVE_FATAL => -30,
  ARCHIVE_FILTER_BZIP2 => 2,
  ARCHIVE_FILTER_COMPRESS => 3,
  ARCHIVE_FILTER_GRZIP => 12,
  ARCHIVE_FILTER_GZIP => 1,
  ARCHIVE_FILTER_LRZIP => 10,
  ARCHIVE_FILTER_LZIP => 9,
  ARCHIVE_FILTER_LZMA => 5,
  ARCHIVE_FILTER_LZOP => 11,
  ARCHIVE_FILTER_NONE => 0,
  ARCHIVE_FILTER_PROGRAM => 4,
  ARCHIVE_FILTER_RPM => 8,
  ARCHIVE_FILTER_UU => 7,
  ARCHIVE_FILTER_XZ => 6,
  ARCHIVE_FORMAT_7ZIP => 917504,
  ARCHIVE_FORMAT_AR => 458752,
  ARCHIVE_FORMAT_AR_BSD => 458754,
  ARCHIVE_FORMAT_AR_GNU => 458753,
  ARCHIVE_FORMAT_BASE_MASK => 16711680,
  ARCHIVE_FORMAT_CAB => 786432,
  ARCHIVE_FORMAT_CPIO => 65536,
  ARCHIVE_FORMAT_CPIO_AFIO_LARGE => 65542,
  ARCHIVE_FORMAT_CPIO_BIN_BE => 65539,
  ARCHIVE_FORMAT_CPIO_BIN_LE => 65538,
  ARCHIVE_FORMAT_CPIO_POSIX => 65537,
  ARCHIVE_FORMAT_CPIO_SVR4_CRC => 65541,
  ARCHIVE_FORMAT_CPIO_SVR4_NOCRC => 65540,
  ARCHIVE_FORMAT_EMPTY => 393216,
  ARCHIVE_FORMAT_ISO9660 => 262144,
  ARCHIVE_FORMAT_ISO9660_ROCKRIDGE => 262145,
  ARCHIVE_FORMAT_LHA => 720896,
  ARCHIVE_FORMAT_MTREE => 524288,
  ARCHIVE_FORMAT_RAR => 851968,
  ARCHIVE_FORMAT_RAW => 589824,
  ARCHIVE_FORMAT_SHAR => 131072,
  ARCHIVE_FORMAT_SHAR_BASE => 131073,
  ARCHIVE_FORMAT_SHAR_DUMP => 131074,
  ARCHIVE_FORMAT_TAR => 196608,
  ARCHIVE_FORMAT_TAR_GNUTAR => 196612,
  ARCHIVE_FORMAT_TAR_PAX_INTERCHANGE => 196610,
  ARCHIVE_FORMAT_TAR_PAX_RESTRICTED => 196611,
  ARCHIVE_FORMAT_TAR_USTAR => 196609,
  ARCHIVE_FORMAT_XAR => 655360,
  ARCHIVE_FORMAT_ZIP => 327680,
  ARCHIVE_MATCH_CTIME => 512,
  ARCHIVE_MATCH_EQUAL => 16,
  ARCHIVE_MATCH_MTIME => 256,
  ARCHIVE_MATCH_NEWER => 1,
  ARCHIVE_MATCH_OLDER => 2,
  ARCHIVE_OK => 0,
  ARCHIVE_READDISK_HONOR_NODUMP => 2,
  ARCHIVE_READDISK_MAC_COPYFILE => 4,
  ARCHIVE_READDISK_NO_TRAVERSE_MOUNTS => 8,
  ARCHIVE_READDISK_RESTORE_ATIME => 1,
  ARCHIVE_RETRY => -10,
  ARCHIVE_VERSION_NUMBER => 3001002,
  ARCHIVE_WARN => -20,
};

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Archive::Libarchive::FFI::Constant

=head1 VERSION

version 0.0802

=head1 SYNOPSIS

 use Archive::Libarchive::FFI qw( :const );

=head1 DESCRIPTION

This document contains a list of all constants provided by the libarchive perl
bindings.  Depending on the version of libarchive that you are using, some
constants may not be available.  You can detect this at run time by using the
can method on the L<Archive::Libarchive::FFI> package, like so:

 if(Archive::Libarchive::FFI->can('ARCHIVE_OK'))
 {
   ...
 }

If provided by your libarchive library, these constants will be available and
exportable from L<Archive::Libarchive::FFI> (you may import all available
constants using the C<:const> export tag).

=head1 CONSTANTS

=over 4

=item AE_IFBLK

=item AE_IFCHR

=item AE_IFDIR

=item AE_IFIFO

=item AE_IFLNK

=item AE_IFMT

=item AE_IFREG

=item AE_IFSOCK

=item ARCHIVE_API_FEATURE

=item ARCHIVE_API_VERSION

=item ARCHIVE_BYTES_PER_RECORD

=item ARCHIVE_COMPRESSION_BZIP2

=item ARCHIVE_COMPRESSION_COMPRESS

=item ARCHIVE_COMPRESSION_GZIP

=item ARCHIVE_COMPRESSION_LRZIP

=item ARCHIVE_COMPRESSION_LZIP

=item ARCHIVE_COMPRESSION_LZMA

=item ARCHIVE_COMPRESSION_NONE

=item ARCHIVE_COMPRESSION_PROGRAM

=item ARCHIVE_COMPRESSION_RPM

=item ARCHIVE_COMPRESSION_UU

=item ARCHIVE_COMPRESSION_XZ

=item ARCHIVE_DEFAULT_BYTES_PER_BLOCK

=item ARCHIVE_ENTRY_ACL_ADD_FILE

=item ARCHIVE_ENTRY_ACL_ADD_SUBDIRECTORY

=item ARCHIVE_ENTRY_ACL_APPEND_DATA

=item ARCHIVE_ENTRY_ACL_DELETE

=item ARCHIVE_ENTRY_ACL_DELETE_CHILD

=item ARCHIVE_ENTRY_ACL_ENTRY_DIRECTORY_INHERIT

=item ARCHIVE_ENTRY_ACL_ENTRY_FAILED_ACCESS

=item ARCHIVE_ENTRY_ACL_ENTRY_FILE_INHERIT

=item ARCHIVE_ENTRY_ACL_ENTRY_INHERIT_ONLY

=item ARCHIVE_ENTRY_ACL_ENTRY_NO_PROPAGATE_INHERIT

=item ARCHIVE_ENTRY_ACL_ENTRY_SUCCESSFUL_ACCESS

=item ARCHIVE_ENTRY_ACL_EVERYONE

=item ARCHIVE_ENTRY_ACL_EXECUTE

=item ARCHIVE_ENTRY_ACL_GROUP

=item ARCHIVE_ENTRY_ACL_GROUP_OBJ

=item ARCHIVE_ENTRY_ACL_INHERITANCE_NFS4

=item ARCHIVE_ENTRY_ACL_LIST_DIRECTORY

=item ARCHIVE_ENTRY_ACL_MASK

=item ARCHIVE_ENTRY_ACL_OTHER

=item ARCHIVE_ENTRY_ACL_PERMS_NFS4

=item ARCHIVE_ENTRY_ACL_PERMS_POSIX1E

=item ARCHIVE_ENTRY_ACL_READ

=item ARCHIVE_ENTRY_ACL_READ_ACL

=item ARCHIVE_ENTRY_ACL_READ_ATTRIBUTES

=item ARCHIVE_ENTRY_ACL_READ_DATA

=item ARCHIVE_ENTRY_ACL_READ_NAMED_ATTRS

=item ARCHIVE_ENTRY_ACL_STYLE_EXTRA_ID

=item ARCHIVE_ENTRY_ACL_STYLE_MARK_DEFAULT

=item ARCHIVE_ENTRY_ACL_SYNCHRONIZE

=item ARCHIVE_ENTRY_ACL_TYPE_ACCESS

=item ARCHIVE_ENTRY_ACL_TYPE_ALARM

=item ARCHIVE_ENTRY_ACL_TYPE_ALLOW

=item ARCHIVE_ENTRY_ACL_TYPE_AUDIT

=item ARCHIVE_ENTRY_ACL_TYPE_DEFAULT

=item ARCHIVE_ENTRY_ACL_TYPE_DENY

=item ARCHIVE_ENTRY_ACL_TYPE_NFS4

=item ARCHIVE_ENTRY_ACL_TYPE_POSIX1E

=item ARCHIVE_ENTRY_ACL_USER

=item ARCHIVE_ENTRY_ACL_USER_OBJ

=item ARCHIVE_ENTRY_ACL_WRITE

=item ARCHIVE_ENTRY_ACL_WRITE_ACL

=item ARCHIVE_ENTRY_ACL_WRITE_ATTRIBUTES

=item ARCHIVE_ENTRY_ACL_WRITE_DATA

=item ARCHIVE_ENTRY_ACL_WRITE_NAMED_ATTRS

=item ARCHIVE_ENTRY_ACL_WRITE_OWNER

=item ARCHIVE_EOF

=item ARCHIVE_EXTRACT_ACL

=item ARCHIVE_EXTRACT_FFLAGS

=item ARCHIVE_EXTRACT_HFS_COMPRESSION_FORCED

=item ARCHIVE_EXTRACT_MAC_METADATA

=item ARCHIVE_EXTRACT_NO_AUTODIR

=item ARCHIVE_EXTRACT_NO_HFS_COMPRESSION

=item ARCHIVE_EXTRACT_NO_OVERWRITE

=item ARCHIVE_EXTRACT_NO_OVERWRITE_NEWER

=item ARCHIVE_EXTRACT_OWNER

=item ARCHIVE_EXTRACT_PERM

=item ARCHIVE_EXTRACT_SECURE_NODOTDOT

=item ARCHIVE_EXTRACT_SECURE_SYMLINKS

=item ARCHIVE_EXTRACT_SPARSE

=item ARCHIVE_EXTRACT_TIME

=item ARCHIVE_EXTRACT_UNLINK

=item ARCHIVE_EXTRACT_XATTR

=item ARCHIVE_FAILED

=item ARCHIVE_FATAL

=item ARCHIVE_FILTER_BZIP2

=item ARCHIVE_FILTER_COMPRESS

=item ARCHIVE_FILTER_GRZIP

=item ARCHIVE_FILTER_GZIP

=item ARCHIVE_FILTER_LRZIP

=item ARCHIVE_FILTER_LZIP

=item ARCHIVE_FILTER_LZMA

=item ARCHIVE_FILTER_LZOP

=item ARCHIVE_FILTER_NONE

=item ARCHIVE_FILTER_PROGRAM

=item ARCHIVE_FILTER_RPM

=item ARCHIVE_FILTER_UU

=item ARCHIVE_FILTER_XZ

=item ARCHIVE_FORMAT_7ZIP

=item ARCHIVE_FORMAT_AR

=item ARCHIVE_FORMAT_AR_BSD

=item ARCHIVE_FORMAT_AR_GNU

=item ARCHIVE_FORMAT_BASE_MASK

=item ARCHIVE_FORMAT_CAB

=item ARCHIVE_FORMAT_CPIO

=item ARCHIVE_FORMAT_CPIO_AFIO_LARGE

=item ARCHIVE_FORMAT_CPIO_BIN_BE

=item ARCHIVE_FORMAT_CPIO_BIN_LE

=item ARCHIVE_FORMAT_CPIO_POSIX

=item ARCHIVE_FORMAT_CPIO_SVR4_CRC

=item ARCHIVE_FORMAT_CPIO_SVR4_NOCRC

=item ARCHIVE_FORMAT_EMPTY

=item ARCHIVE_FORMAT_ISO9660

=item ARCHIVE_FORMAT_ISO9660_ROCKRIDGE

=item ARCHIVE_FORMAT_LHA

=item ARCHIVE_FORMAT_MTREE

=item ARCHIVE_FORMAT_RAR

=item ARCHIVE_FORMAT_RAW

=item ARCHIVE_FORMAT_SHAR

=item ARCHIVE_FORMAT_SHAR_BASE

=item ARCHIVE_FORMAT_SHAR_DUMP

=item ARCHIVE_FORMAT_TAR

=item ARCHIVE_FORMAT_TAR_GNUTAR

=item ARCHIVE_FORMAT_TAR_PAX_INTERCHANGE

=item ARCHIVE_FORMAT_TAR_PAX_RESTRICTED

=item ARCHIVE_FORMAT_TAR_USTAR

=item ARCHIVE_FORMAT_XAR

=item ARCHIVE_FORMAT_ZIP

=item ARCHIVE_LIBRARY_VERSION

=item ARCHIVE_MATCH_CTIME

=item ARCHIVE_MATCH_EQUAL

=item ARCHIVE_MATCH_MTIME

=item ARCHIVE_MATCH_NEWER

=item ARCHIVE_MATCH_OLDER

=item ARCHIVE_OK

=item ARCHIVE_READDISK_HONOR_NODUMP

=item ARCHIVE_READDISK_MAC_COPYFILE

=item ARCHIVE_READDISK_NO_TRAVERSE_MOUNTS

=item ARCHIVE_READDISK_RESTORE_ATIME

=item ARCHIVE_RETRY

=item ARCHIVE_VERSION_NUMBER

=item ARCHIVE_VERSION_STAMP

=item ARCHIVE_WARN

=back

=head1 SEE ALSO

=over 4

=item L<Archive::Libarchive::FFI>

=item L<Archive::Libarchive::FFI::Callback>

=item L<Archive::Libarchive::FFI::Function>

=back

=head1 AUTHOR

Graham Ollis <plicease@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Graham Ollis.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
